src/mpl.cpp

```diff
@@ -5,187 +5,194 @@
 
     return hal::success();
 }
 
-/*
+/**
 * @brief Set bits in a register without overwriting existing register state
+* @param p_i2c The I2C peripheral used for communication with the device.
 * @param reg_addr: 8 bit register address
 * @param bits_to_set: 8 bit value specifying which bits to set in register
 */
-hal::status mpl::modify_reg_bits(hal::byte reg_addr, hal::byte bits_to_set)
+hal::status modify_reg_bits(hal::i2c* p_i2c, hal::byte reg_addr, hal::byte bits_to_set)
 {
     // Read old register value
-    std::array<hal::byte, 1> reg_payload { reg_addr };
-    std::array<hal::byte, 1> reg_buffer {};
-    HAL_CHECK(hal::write_then_read(*m_i2c, device_address, reg_payload, reg_buffer, hal::never_timeout()));
+    auto reg_buffer = HAL_CHECK(hal::write_then_read<1>(*p_i2c, device_address, std::array<hal::byte, 1>{ reg_addr }, hal::never_timeout()));
 
     // Set specified bits while maintaining old values
-    hal::byte updated_reg = reg_buffer[0] |= bits_to_set;
-    std::array<hal::byte, 2> update_payload { reg_addr, updated_reg };
-    HAL_CHECK(hal::write(*m_i2c, device_address, update_payload, hal::never_timeout()));
+    hal::byte updated_reg = reg_buffer[0] | bits_to_set;
+    HAL_CHECK(hal::write(*p_i2c, device_address, std::array<hal::byte, 2>{ reg_addr, updated_reg }, hal::never_timeout()));
 
     return hal::success();
 }
 
-/*
+/**
 * @brief Wait for a specified flag bit in a register to be set to the desired state.
+* @param p_i2c The I2C peripheral used for communication with the device.
 * @param reg: 8 bit value specifying the register address
 * @param flag: 8 bit value specifying which bit(s) to check
 */
-hal::status mpl::poll_flag(hal::byte reg, hal::byte flag, bool desired_state)
+hal::status poll_flag(hal::i2c* p_i2c, hal::byte reg, hal::byte flag, bool desired_state)
 {   
     std::array<hal::byte, 1> status_payload { reg };
     std::array<hal::byte, 1> status_buffer {};
     bool flag_set = true;
 
     while (flag_set) {
-        hal::result<hal::i2c::transaction_t> res = hal::write_then_read(
-                                                            *m_i2c, 
-                                                            device_address, 
-                                                            status_payload, 
-                                                            status_buffer, 
-                                                            hal::never_timeout());
-        // Only update when no errors (catch device not found)
-        if (res) {
+        if (hal::write_then_read(
+                *p_i2c, 
+                device_address, 
+                status_payload, 
+                status_buffer, 
+                hal::never_timeout())
+        ) {
+            // Only update when no errors (catch device not found)
             flag_set = desired_state ? ((status_buffer[0] & flag) == 0) : ((status_buffer[0] & flag) != 0);
```

Comment:
```suggestion
            // Only update when no errors (catch device not found)
            flag_set = desired_state ? ((status_buffer[0] & flag) == 0) : ((status_buffer[0] & flag) != 0);
```

This section is still quite confusing to read and understand. I feel like desired_state should be equated to the flag bits value being a 1 or a 0. Rename `flag_set` to `finished` and do something like 

```cpp
finished = desired_state == flag_bit_value;
```